WIP - first draft of array map/reduce with zipping

willghatch opened the pull request:
Some questions:

* Right now I've left the code raising the same error as `array_zip` when the array lengths don't match.  Should we make different errors for each, or genericize the zip error?
* Is the ALGO connector code correct?  I did it certainly wrong a couple of times, and I hope this one is at least close.  Your colloquium probably would have been helpful earlier -- eg. explaining what all of the functions with two-letter names mean.  I feel like it always takes me a long time to try to decipher what the intended meanings of the short names are.  Often there's a domino effect where I want to figure out what one abbreviation means, so I go to its definition to find a short body itself only full of abbreviations.  Often I give up.  Ultimately my understanding of what is going on, especially in the ALGO connector, is pretty foggy.
* What are your feelings about how non-duplicated code should be?  There is a loop pattern in the two ALGO connector branches that is similar that could be captured.  I started doing it, but then got a weird parse error that I didn't understand and just backed out my changes and decided to push and get feedback before fussing with it anymore.  Personally, in my own code, that loop pattern is the sort of thing that I'll often leave duplicated if I only do it twice, especially if I can't capture it textually near its two use sites, judging that the extra indirection is not yet worth the benefits of deduplication.  But this is a judgment call that many people and organizations feel differently about, and my impression @jeapostrophe is that you're more of a stickler for deduplication than most.
* I added a couple test files that fail when array lengths aren't matched, and then one that compiles correctly that includes map and reduce.  But what is our general approach to testing that the functionality actually works as expected beyond just compiling correctly?

jeapostrophe commented on hs/src/Reach/AddCounts.hs:
```diff
@@ -91,16 +91,16 @@ instance AC DLStmt where
       f' <- ac f
-      ac_visit $ [x, z]
-      return $ DL_ArrayReduce at ans x z b a i f'
+      ac_visit $ xs ++ [z]
```
You should use `<>` for robustness against the future

jeapostrophe commented on hs/src/Reach/AddCounts.hs:
```diff
@@ -91,16 +91,16 @@ instance AC DLStmt where
+      ac_visit $ xs
+      return $ DL_ArrayMap at ans xs as i f'
+    DL_ArrayReduce at ans xs z b as i f -> do
       -- XXX remove if ans not used
```
You should do this since you're thinking about this code and testing it

jeapostrophe commented on hs/src/Reach/AnalyzeVars.hs:
```diff
@@ -172,8 +172,8 @@ instance FreeVars DLStmt where
     DL_Let _ _ e -> freeVars e
-    DL_ArrayMap _ _ x a i f -> freeVars [x] <> bindsFor [a, i] f
-    DL_ArrayReduce _ _ x z a b i f -> freeVars [x, z] <> bindsFor [a, b, i] f
+    DL_ArrayMap _ _ xs as i f -> freeVars xs <> bindsFor (as ++ [i]) f
```
<>

jeapostrophe commented on hs/src/Reach/Backend/JS.hs:
```diff
@@ -531,24 +531,24 @@ jsCom = \case
     r' <- jsArg r
-    return $ "const" <+> ans' <+> "=" <+> "await" <+> jsApply "stdlib.Array_asyncMap" [ x', (jsApply "async" [a', i'] <+> "=>" <+> jsBraces (f' <> hardline <> jsReturn r')) ]
-  DL_ArrayReduce _ ans x z b a i (DLBlock _ _ f r) -> do
+    return $ "const" <+> ans' <+> "=" <+> "await" <+> jsApply "stdlib.Array_asyncMap" (xs' ++ [(jsApply "async" (as' ++ [i']) <+> "=>" <+> jsBraces (f' <> hardline <> jsReturn r'))])
```
You have to actually implement this extension

jeapostrophe commented on hs/src/Reach/Backend/JS.hs:
```diff
@@ -531,24 +531,24 @@ jsCom = \case
     f' <- jsPLTail f
     r' <- jsArg r
-    return $ "const" <+> ans' <+> "=" <+> "await" <+> jsApply "stdlib.Array_asyncReduce" [ x', z', (jsApply "async" [b', a', i']) <+> "=>" <+> jsBraces (f' <> hardline <> jsReturn r') ]
+    return $ "const" <+> ans' <+> "=" <+> "await" <+> jsApply "stdlib.Array_asyncReduce" (xs' ++ [z', (jsApply "async" $ [b'] ++ as' ++ [i']) <+> "=>" <+> jsBraces (f' <> hardline <> jsReturn r')])
```
And this one

jeapostrophe commented on hs/src/Reach/CollectCounts.hs:
```diff
@@ -213,10 +213,10 @@ instance CountableK DLStmt where
-    DL_ArrayReduce _ ans x z b a i f ->
-      count_rms [ans, b, a, i] (counts f <> kcs) <> counts [x, z]
+    DL_ArrayMap _ ans xs as i f ->
+      count_rms (as ++ [ans, i]) (counts f <> kcs) <> counts xs
```
<>

jeapostrophe commented on hs/src/Reach/Connector/ALGO.hs:
```diff
@@ -1938,34 +1938,46 @@ cm km = \case
+                  store_let iv True load_idx $
+                  arraysLoop xsTail asTail
+              (_, _) -> impossible "DL_ArrayMap: inconsistent array sizes"
+        arraysLoop xs as
```
I would rewrite this to be like `foldlM xyz go $ zip xs as` where `xyz` is the `cp ... body` part and `go` is basically the double cons case

jeapostrophe commented on hs/src/Reach/Connector/ALGO.hs:
```diff
@@ -1938,34 +1938,46 @@ cm km = \case
+                    store_let iv True load_idx $ do
+                      arraysLoop xsTail asTail
+                (_, _) -> impossible "DL_ArrayReduce: inconsistent array sizes"
+          arraysLoop xs as
```
Ditto

jeapostrophe commented on hs/src/Reach/Connector/ETH_Solidity.hs:
```diff
@@ -994,12 +994,14 @@ solCom = \case
-    addMemVars $ [ans, a]
-    let sz = arraySize x
+  DL_ArrayMap _ ans xs as i (DLBlock _ _ f r) -> do
+    addMemVars $ [ans] ++ as
```
<>

jeapostrophe commented on hs/src/Reach/Connector/ETH_Solidity.hs:
```diff
@@ -994,12 +994,14 @@ solCom = \case
-    let sz = arraySize x
+  DL_ArrayMap _ ans xs as i (DLBlock _ _ f r) -> do
+    addMemVars $ [ans] ++ as
+    let sz = case allEqual $ map arraySize xs of
```
You should make a function for this... `arraySizes` and stick it in the same place as `arraySize`

jeapostrophe commented on hs/src/Reach/EPP.hs:
```diff
@@ -359,20 +359,20 @@ be_m = \case
-    fg_defn $ [ans, a, i]
-    fg_use $ x
+  DL_ArrayMap at ans xs as i f -> do
+    fg_defn $ as ++ [ans, i]
```
<>

jeapostrophe commented on hs/src/Reach/Optimize.hs:
```diff
@@ -620,19 +620,23 @@ instance Optimize DLStmt where
-          _ -> def
+      maybeUnroll :: DLStmt -> [DLArg] -> App DLStmt -> App DLStmt
+      maybeUnroll s xs def =
+        let arrayLengths = map (\x -> case argTypeOf x of
```
Never use `let ... in`, make a `do` block and use a `let` that doesn't indent

jeapostrophe replied:
Also, this looks like a use of `arraySizes`

jeapostrophe commented on hs/src/Reach/UnrollLoops.hs:
```diff
@@ -119,15 +120,15 @@ instance Unroll DLStmt where
-      r' <- zipWithM (\xa iv -> fu_ fb [(a, xa), (i, (DLA_Literal $ DLL_Int at iv))]) x' [0..]
+    DL_ArrayMap at ans xs as i fb -> do
+      (_, xs') <- unzip <$> mapM (ul_explode at) xs
+      r' <- zipWithM (\xa iv -> fu_ fb $ (zip as xa) ++ [(i, (DLA_Literal $ DLL_Int at iv))]) (transpose xs') [0..]
```
<>

jeapostrophe commented on hs/src/Reach/Util.hs:
```diff
@@ -156,3 +157,10 @@ makeErrCode errType errIndex =
 arraySet :: Int -> a -> [a] -> [a]
 arraySet n a arr = take n arr <> [a] <> drop (n + 1) arr
+
+allEqual :: Eq a => [a] -> Either (Maybe (a,a)) a
```
I believe there is a similar function called `allSame` somewhere

jeapostrophe commented on hs/t/n/array_map_length_mismatch.rsh:
```diff
@@ -0,0 +1,20 @@
+
+export const main =
+  Reach.App(
+    {},
```
Never use the ternary style in new code

jeapostrophe commented on hs/t/n/array_map_length_mismatch.rsh:
```diff
@@ -0,0 +1,20 @@
+'reach 0.1';
```
You need to add the `.txt` files that `make hs-test-accept` generates

jeapostrophe commented:
> * Right now I've left the code raising the same error as `array_zip` when the array lengths don't match.  Should we make different errors for each, or genericize the zip error?

`impossible` errors should never be seen by users. Other errors, it is fine to say "The two arrays don't match in size"

> * Is the ALGO connector code correct?  I did it certainly wrong a couple of times, and I hope this one is at least close.  Your colloquium probably would have been helpful earlier -- eg. explaining what all of the functions with two-letter names mean.  I feel like it always takes me a long time to try to decipher what the intended meanings of the short names are.  Often there's a domino effect where I want to figure out what one abbreviation means, so I go to its definition to find a short body itself only full of abbreviations.  Often I give up.  Ultimately my understanding of what is going on, especially in the ALGO connector, is pretty foggy.

You should write a few examples (`examples` get run, whereas `hs/t` just get compiled) that check that the result is the right thing. It looks roughly okay to me

> * What are your feelings about how non-duplicated code should be?  There is a loop pattern in the two ALGO connector branches that is similar that could be captured.  I started doing it, but then got a weird parse error that I didn't understand and just backed out my changes and decided to push and get feedback before fussing with it anymore.  Personally, in my own code, that loop pattern is the sort of thing that I'll often leave duplicated if I only do it twice, especially if I can't capture it textually near its two use sites, judging that the extra indirection is not yet worth the benefits of deduplication.  But this is a judgment call that many people and organizations feel differently about, and my impression @jeapostrophe is that you're more of a stickler for deduplication than most.

Yea, I'm a pretty big de-duplication stickler. I might for instance make a `forArrays` helper that can handle both what `ArrayMap` and `ArrayReduce` do.

> * I added a couple test files that fail when array lengths aren't matched, and then one that compiles correctly that includes map and reduce.  But what is our general approach to testing that the functionality actually works as expected beyond just compiling correctly?

You should make compile-only tests that use the theorem prover to assert that an answer is the right thing. And you should make example programs that compute something and check that the JS reads the right answer from the blockchain.

Also, you should remove `ArrayZip`

willghatch commented:
Here is a new version.  Some comments:

* The ALGO backend code seems to work.  But I'm not comfortable enough with Haskell's lazy semantics to be 100% confident that it's really doing what I think it is.
* There are a couple of new verification errors, and I'm not sure what's wrong or how to approach fixing things.

The first verification error is in `y/no-tail-if`:
```
user error (Expected success, but failed.
      Verifying knowledge assertions
      Verifying for generic connector
        Verifying when ALL participants are honest
        Verifying when NO participants are honest
      Checked 17 theorems; No failures!
      reachc: error[RX0000]: Cannot inspect value from `forall`
      
        ./no-tail-if.rsh:18:7:application
      
        18| forall(UInt, (countA) =>
      
      For further explanation of this error, see: https://docs.reach.sh/rsh/errors/#RX0000
      
      )
```


The second verification error is in `y/overflows.rsh` (and is long):
```
user error (Expected success, but failed.
      Verifying knowledge assertions
      Verifying for generic connector
        Verifying when ALL participants are honest
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 39;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [39, 39, 38, 37, 15, 37, 1, 1, 12, 13 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v108 = v64[0] <= (UInt.max - 1);
        //    ^ would be false
        assert(v108);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 39;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [36, 39, 38, 37, 15, 37, 1, 1, 12, 13 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v113 = v64[1] <= (UInt.max - 1);
        //    ^ would be false
        assert(v113);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 21240;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [0, 1, 21240, 2437, 8855, 11797, 1142, 281, 15921, 9371, 5853 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v118 = v64[2] <= (UInt.max - 1);
        //    ^ would be false
        assert(v118);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 2440;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [2, 0, 1, 2440, 1532, 2033, 1142, 281, 1275, 1320, 971 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v123 = v64[3] <= (UInt.max - 1);
        //    ^ would be false
        assert(v123);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 8859;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [2, 3, 2, 1, 8859, 2937, 8365, 5705, 1590, 4032, 5853, 281, 7061 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v128 = v64[4] <= (UInt.max - 1);
        //    ^ would be false
        assert(v128);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 11802;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [3, 4, 3, 2, 0, 11802, 8365, 8679, 10450, 7006, 5853, 281, 4118 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v133 = v64[5] <= (UInt.max - 1);
        //    ^ would be false
        assert(v133);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 8371;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [2, 3, 4, 1, 5, 1, 8371, 7169, 2078, 5496, 5853, 281, 7549 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v138 = v64[6] <= (UInt.max - 1);
        //    ^ would be false
        assert(v138);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 32292;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [3, 5, 0, 1, 0, 2, 0, 32292, 10450, 30612, 5853, 281, 15921 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v143 = v64[7] <= (UInt.max - 1);
        //    ^ would be false
        assert(v143);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 10458;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [6, 7, 6, 7182, 7, 1142, 281, 5462, 9694, 5853 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v148 = v64[8] <= (UInt.max - 1);
        //    ^ would be false
        assert(v148);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 30621;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [2, 6, 7, 7, 8, 7, 4, 2, 1, 30621, 5853, 281, 15921 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v153 = v64[9] <= (UInt.max - 1);
        //    ^ would be false
        assert(v153);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 5863;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [8, 9, 8, 7, 0, 7, 281, 4193, 2, 5863 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v158 = v64[10] <= (UInt.max - 1);
        //    ^ would be false
        assert(v158);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 28111;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [5, 8, 8, 6, 0, 5, 5, 0, 0, 10, 1, 281, 15921 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v163 = v64[11] <= (UInt.max - 1);
        //    ^ would be false
        assert(v163);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 1154;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [2, 3, 10, 1, 11, 1, 1, 5, 10, 0, 9, 281, 906 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v168 = v64[12] <= (UInt.max - 1);
        //    ^ would be false
        assert(v168);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 294;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [12, 10, 9, 7, 2, 11, 294, 286, 10, 3 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v173 = v64[13] <= (UInt.max - 1);
        //    ^ would be false
        assert(v173);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 20551;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [10, 5, 0, 20551, 7, 9, 8, 15921, 8 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v178 = v64[14] <= (UInt.max - 1);
        //    ^ would be false
        assert(v178);
      
      Verification failed:
        when ALL participants are honest
        of theorem: assert
        msg: "add overflow"
        at ./overflows.rsh:14:30:application
        at ./overflows.rsh:14:20:application call to [unknown function] (defined at: ./overflows.rsh:14:24:function exp)
      
        // Violation Witness
      
        const UInt.max = 15936;
      
        const v64 = protect<Array(UInt, 16)>("A".interact.get());
        //    ^ could = array(UInt, [8, 9, 13, 7, 10, 7, 11, 15936, 12, 3 ] )
        //      from: ./overflows.rsh:9:40:application
      
        // Theorem Formalization
      
        const v183 = v64[15] <= (UInt.max - 1);
        //    ^ would be false
        assert(v183);
      
        Verifying when NO participants are honest
      Checked 36 theorems; 32 failures (and 16 omitted repeats) :'(
      )
```

jeapostrophe commented on docs/src/rsh/compute/index.md:
```diff
@@ -1025,7 +1026,7 @@ arr.mapWithIndex(f)
  `{!rsh} Array.mapWithIndex(arr, f)` is similar to `{!rsh} Array.map`, except it
 provides `{!rsh} f` with an additional argument, which is the index of the current element in `{!rsh} arr`.
-Unlike `{!rsh} Array.map`, this function is not generalized to an arbitrary number of arrays; it only accepts one array.
+The index argument must be the last argument of the given function `{!rsh} f`.
```
I don't think you mean "must be". I think you just mean "is"

jeapostrophe commented on docs/src/rsh/compute/index.md:
```diff
@@ -1067,7 +1071,7 @@ arr.reduceWithIndex(z, f)
  `{!rsh} Array.reduceWithIndex(arr, z, f)` is similar to `{!rsh} Array.reduce`, except it
 provides `{!rsh} f` with an additional argument, which is the index of the current element in `{!rsh} arr`.
-Unlike `{!rsh} Array.reduce`, this function is not generalized to an arbitrary number of arrays; it only accepts one array.
+The index argument must be the last argument of the given function `{!rsh} f`.
```
Ditto

jeapostrophe commented on hs/src/Reach/AST/DLBase.hs:
```diff
@@ -1165,3 +1159,13 @@ type DLViews = InterfaceLikeMap IType
+arraysLength :: [DLArg] -> Integer
+arraysLength arrays = do
+  let sizes = map (\x -> case argTypeOf x of
+                                 T_Array _ n -> Just n
```
There's already a function that does this

jeapostrophe replied:
Or if I'm remember wrong... there should be be :)

jeapostrophe replied:
`arrTypeLen` is what I'm thinking of. You don't care about the `Just` because you're just going to do an impossible later

willghatch commented:
And with respect to the `allTheSame` function, I replaced its uses with uses of the `allEqual` that I wrote and removed `allTheSame`.  They did the same thing but mine provides witness values for inequality that are nice for error messages.

jeapostrophe commented on hs/src/Reach/Connector/ALGO.hs:
```diff
@@ -1938,34 +1923,42 @@ cm km = \case
+               store_let iv True load_idx $
+               k
+        let bigK = F.foldrM bodyF finalK $ zip xs as
+        bigK ()
```
Just put in the expression and drop the ()

jeapostrophe commented on hs/src/Reach/Connector/ALGO.hs:
```diff
@@ -1938,34 +1923,42 @@ cm km = \case
+                 store_let iv True load_idx $
+                 k
+          let bigK = F.foldrM bodyF finalK $ zip xs as
+          bigK ()
```
Ditto

jeapostrophe commented on hs/src/Reach/Optimize.hs:
```diff
@@ -447,16 +446,7 @@ class Extract a where
 
-allTheSame :: (Eq a, Sanitize a) => [a] -> Maybe a
-allTheSame xs =
-  case map sani xs of
```
Notice that my version uses `Sanitize` to remove SrcLocs and that makes it more likely things will be equal. If you want to replace `allTheSame`, then either you should use `Sanitize` too or you should do `allTheSame = allEqual . map sani`

jeapostrophe commented on hs/t/n/array_map_length_mismatch.txt:
```diff
@@ -0,0 +1,11 @@
+reachc: error[RE0071]: Zip requires arrays of equal length, but got 5 and 3
```
Change the text to not reference zip

jeapostrophe commented on hs/t/y/array_groups.txt:
```diff
@@ -4,6 +4,6 @@ Verifying for generic connector
 WARNING: Compiler instructed to emit for Algorand, but the conservative analysis found these potential problems:
- * This program could use 34445 units of cost, but the limit is 700; longest path:
-     TOP --> preamble --> api --> publish --> l0_publish_lt_1 --> publish0 --> l2_checkTxnK --> l3_forTop --> l4_forEnd --> l5_forTop --> l7_forTop --> l8_forEnd --> l6_forEnd --> l9_forTop --> l10_forEnd --> updateState --> apiReturn --> checkSize --> done --> BOT
+ * This program could use 2665 units of cost, but the limit is 700; longest path:
```
Woah! check that out!

willghatch commented:
I'm not sure why I didn't notice the `sani` bit.  Sorry about that.

I've fixed things you noticed, except that I haven't quite figured out how to remove the `()` and make the type system happy yet.

That said, I still have the verification errors.  Can you give me some advice for fixing them?

Anyway, I'm going to go eat some lunch now.

jeapostrophe commented on hs/src/Reach/Connector/ALGO.hs:
```diff
@@ -1938,34 +1923,40 @@ cm km = \case
+               sallocLet x (return ()) $
+                 store_let iv True load_idx $
+                 k
+          (F.foldrM bodyF finalK $ zip xs as) ()
```
You definitely don't need these `()`s... what error do you get? (You do need to remove both of them)

willghatch replied:
Removing them gives me this error:

```
/home/wgh/s/mk/reach-lang/hs/src/Reach/Connector/ALGO.hs:1942:19: error:
    • Couldn't match type ‘()’ with ‘ReaderT Env IO ()’
      Expected type: (DLVar, DLArg) -> App () -> ReaderT Env IO (App ())
        Actual type: (DLVar, DLArg)
                     -> App (App ()) -> ReaderT Env IO (App ())
    • In the first argument of ‘F.foldrM’, namely ‘bodyF’
      In the expression: F.foldrM bodyF finalK
      In a stmt of a 'do' block: (F.foldrM bodyF finalK $ zip xs as)
     |
1942 |         (F.foldrM bodyF finalK $ zip xs as)
     |                   ^^^^^
```

Maybe there's an obvious next step, but I'm not sure what it is.

jeapostrophe reviewed (changes requested):
I am about to study the verification errors

jeapostrophe commented on hs/src/Reach/UnrollLoops.hs:
```diff
@@ -119,15 +113,15 @@ instance Unroll DLStmt where
-      r' <- zipWithM (\xa iv -> fu_ fb [(a, xa), (i, (DLA_Literal $ DLL_Int at iv))]) x' [0..]
+    DL_ArrayMap at ans xs as i fb -> do
+      (_, xs') <- unzip <$> mapM (ul_explode at) xs
+      r' <- zipWithM (\xa iv -> fu_ fb $ (zip as xa) <> [(i, (DLA_Literal $ DLL_Int at iv))]) (transpose xs') [0..]
```
That's what `fu_` does ... "freshen up"... that's our name for alpha-varying. Basically because we're unrolling, we need to make new variables for the body of the loop because it will occur many times now

willghatch commented:
Aha!  I finally figured it out.  Well, at least partially.

The `forEach` that has the `assume` is compiled into an array_reduce, and I finally figured out that the problem is that it's being removed by the `AddCounts` optimization.  IE the result of the reduce is never used!  I'm not sure why the other `forEach` that includes the `require` isn't compiled away too, though.

jeapostrophe commented:
Ah, it shouldn't remove it if the body isn't pure (look at the similar code for Let)

jeapostrophe reviewed (commented):
.

willghatch commented:
I believe it all works now.  One thing I would maybe change now though is to do the `isPure` test on the whole statement in `addCounts` to not duplicate the implementation if `isPure` for map, reduce, and let.

jeapostrophe reviewed (approved):
Approve, assuming CI passes

willghatch commented:
The overflow test apparently no longer requires more than 700 cost.  I added that, assuming it is correct.
